### DISCUSSION TITLE
Tests not firing in the remote

### DISCUSSION THREAD

**Discussion by blackfalcon** [AlaskaAirlines/WC-Generator#523] [Announcements]
## The problem

@jason-capsule42  merged a PR https://github.com/AlaskaAirlines/auro-dropdown/pull/217 and rained questions about why this release failed. https://github.com/AlaskaAirlines/auro-dropdown/actions/runs/8161376207
 
There are a couple of things here. One is the new a11y linting error. I believe that this was caused by a dependency update. Which one, not sure. Still looking into that. 
 
This is what put a spotlight on the issue. What is weird is that the Node 18 and 20 tests didn't fail, but the release did. Why? Also, why did this not fail locally on Jason's machine when a pre-commit should have caught that? 
 
re: Jason's local, I can only assume that with the recent dependency updates his local was out of sync. That will happen. Also, understand that this new error is not a recent code change, the error is appearing because a linting dependency update now sees it. Yay, updated tools finding years-old errors. This is how we get better. 
 
What this REALLY surfaced is another interesting fact. Why did this ONLY fail on release? This gets a little deep, I will do my best here. 
 
Recently I made some updates to our repos that will post back current released versions into the documentation. In order to do this I need to build the docs in the remote. Pretty straightforward. 
 
To trigger this, there is a new command in the ./package.json called `postversion`. What this does is, in the same action as the semantic-release process, a build and a commit are happening in the remote (Github). This happens here https://github.com/AlaskaAirlines/auro-dropdown/actions/runs/8161376207/job/22310029584#step:6:228
 
Like I said before, the pre-commit that usually happens locally to validate a commit is now also happening in the remote. And if you have the updated dependencies, you will get the same error. 
 
**Conclusion**: I can't say that this is happening in all repos, but it's happening. Tests and linters MUST run in the remote with the Node 18/20 builds. It's great that we are getting successful builds, but not running tests, that's bad. That must be fixed. 
 
What I feel happened is that test and linters were removed from the build script to keep local builds fast. I get that. But with revisions to the GitHub action workflow over time, we forgot about that change and never maintained those actions. 
 
## Recommentation

The following options are on the table. 
 
1. Add test and linters back into the build script and accept that doing a local build will be a little longer
1. Create a new build:release script that has the missing parts and ensure that runs with pre-commits AND update the workflow to use that versus the standard build we have today.
1. Declaratively add tests and linters to the GitHub actions workflow. 
 
I welcome your feedback.

**Comment by blackfalcon** [comment]
@jason-capsule42 replied 

Thanks for summarizing here Dale.
 
Regarding your points at the end:
 
1.  I 100% agree that the local build should be running test and linters. I think it may be useful to explore having another tier of linting happening live while npm run dev is going but with more leniency (e.g. don't complain about extra lines, commented out code, console.log(ing), JS Doc syntax.... just things that are actually how the actual lines JS/CSS is actually written so that we don't write up a bunch of new code and then have to refactor a bunch of it later. We can specify this by having a script like eslint-dev which passes in a lint config file that just imports our strict one but then over rides some things that are commanly used intentionally during development like I listed earlier. 
2. Regarding this - are you saying that this is an option in lieu of adding  the test and linters script back into the main build? If we did that when/why would we ever run build rather than build:release?
3. Again is this a third option or an "also do this..."?

**Reply by blackfalcon** [reply]
[Tuesday 4:03 PM] Dale Sande

I am hesitant to put test and linters in the build again as this will add time to that action when running a build locally for whatever reason is needed. 
 
When I come to a repo I haven't touched in a while, I will `git pull`, then `npm ci` to ensure dependencies are up to date per the pull. I will then run a `npm run build:ci` to ensure that the repo's build resources are correct and stable. I am not concerned about tests linters or bundled resources at this time. 
 
The `build:release` would be all the things happening and I would assume that they would take a little longer. With this we can also look at adding other things for code health that are time-consuming and not something we want to deal with all the time. prettier? 
 
We could also look at a `build:pre-commit` as well. Pre-commits are great but also annoying. They MUST be fast. So what are the minimum things we want in a pre-commit test? Do we need the build? Those resources should be intact from dev. Do we need bundler? Tests and linters, sure. Things can fail on the remote, it's just a matter of where we want to deal with that. With every commit or find out that the remote build failed? 
 
I like the idea of a build:release then we could move all of the time-consuming tooling to that, like test linters and bundler. 
 
re: live tests. We already support this with each repo. `npm run test:watch`
 
re: linter watch, there are things that support that, we would have to look into that. 

All of these are ideal situations. My only thing and why I am bringing this discussion here is that this is something we NEED to address. And since I will have to go in and touch every repo, it would be nice if we had a consensus on a direction versus a knee-jerk response.

**Reply by blackfalcon** [reply]
[Tuesday 4:09 PM] @jason-capsule42 

Well, let's talk about the distinction between what is strictly necessary to prevent the issue that happened today with Dropdown vs. the nice to haves.
 
Is it correct that simply adding `build:release` and having the pre-commit and GitHub actions used would prevent this issue from happening again?

**Reply by blackfalcon** [reply]
No. Nothing can really do that. I am suggesting a path where are are not adding more time to the pre-commits.

**Comment by blackfalcon** [comment]
re: release issue

The failed release https://github.com/AlaskaAirlines/auro-dropdown/actions/runs/8161376207. With additional research, I discovered that the issue was caused by the following commit. It appears that @irma-kurnia-alaska had removed the lining bypass rules. 

- https://github.com/AlaskaAirlines/auro-dropdown/commit/d1db5bca8559065dce1f90c7197b40341ef7ec13#diff-ab7bd51ef198950a30a14340022076364f8c8c4701983f13e6a0ed1f0d507691

I have created the issue https://github.com/AlaskaAirlines/auro-dropdown/issues/222 to solve.

Please note that this update was done with this commit https://github.com/AlaskaAirlines/auro-dropdown/commit/7039e379fd38c81c90676bf2793525b44d287d0a

Which was also updated with this commit https://github.com/AlaskaAirlines/auro-dropdown/commit/a7aed9cccea9422f4b5d4d72103a7c5478dbe075

Which was originally added with this commit https://github.com/AlaskaAirlines/auro-dropdown/commit/13ec34e837c90da0c2d288f98d15572dbb292d55

I am unsure why this listing bypass was added, removed, added back, and then removed again. What concerns me is that a listing bypass for an accessibility warning was added and there is no linked issue or body in the commit message describing why this was bypassed versus addressed. This is requested with issue https://github.com/AlaskaAirlines/auro-dropdown/issues/222

**Reply by fajar-apri-alaska** [reply]
Hi Team,
I just want to say my part here.

I have mostly read this discussion, and also follow up with @irma-kurnia-alaska what's really happening there.
Confirmed that she just removed the bypass rules because after doing `rebase` against the main branch, it is became a blocker on the build step (to commit/pre-commit) in Her local (because that the rules definition is not there anymore, or something like that), and yes the eslint error is also showing on my local machine.

<img width="928" alt="image" src="https://github.com/AlaskaAirlines/WC-Generator/assets/124668226/8e75eb45-8341-4565-be66-51c27a37d431">

I also have the same concern with what Dale said in the last part.

>I am unsure why this listing bypass was added, removed, added back, and then removed again. What concerns me is that a listing bypass for an accessibility warning was added and there is no linked issue or body in the commit message describing why this was bypassed versus addressed.

Because I also said to her **when the error happens, and before I found about this discussion**: "Maybe there is a time in the past where that bypass rules are needed to make things fast or something, but we (me and Irma) should just ignore the reason. Instead, this time we(me and Irma) should fix that so the code became standard (could pass the linters)."

But anyway, The point of this discussion is to address why the `test` is not firing and an action so that it wouldn't happened again , so I will follow any direction that the team is going with.

Thanks before.

**Comment by blackfalcon** [comment]
re: tests in the remote not firing

## The problem

With this commit, `linters` and `test` were part of the `build` script https://github.com/AlaskaAirlines/WC-Generator/commit/1b1ac547cac8333acd1c6a957714662d83cd2d70

The `test` script was removed from the `build` script https://github.com/AlaskaAirlines/WC-Generator/commit/a2d48e0f95777f2f9d448f67deef2867eed8aaf1#diff-4d62358b6f4b985933a16fc493562e0c877aa83736213879215b48bb02cdda41

The `linters` script was removed in 2021 https://github.com/AlaskaAirlines/WC-Generator/commit/0cbb79992095c3d261937982a491276ed9663aa0. 

Both of these were removed with the intention of making the `build` process faster and pushing code testing and linting to only the pre-commit phase. The standard for this is: 

```shell
WC-Generator/template/.husky/pre-commit.temp
4:./node_modules/.bin/npm-run-all preCommit test linters postinstall
```  

The HUGE miss here was that when these were removed, there was no backup plan where linters were running in the remote. This became 100% dependent on local developers to always run the pre-commit. 

All repos have been updated to support `"postversion"` which runs the `build:api` and `build:docs` scripts which then triggers a pre-release in the remote. All repos have this. This ONLY happens in the `release` pipeline that that's why the previous Node version tests are not picking up `test` or `linters` issues. 

## Proposal

I am still in favor of running a speedy build when needed. Adding linters and tests will effectively slow things down. 

Developers would use `npm run build` for local building of assets. This is lightweight and fast. Notice that none of the docs building scripts are included, except for `build:release`. It's considerable to think that these resources will be updated via development. Regardless the doc generators will be run in the remote and committed back to the repo with the release. 

```json
"build": "npm-run-all build:sass sass:render types",
```

I propose a new script, `build:test`.  A developer could run this locally, but it's intended to be used with the `testPublish.yml` config to test builds for both Node versions. 

```json
"build:test": "npm-run-all test linters",
```

I propose a new script, `build:release`. Again, a developer would never need to use this locally. It's intended to build out all the assets required for the release. 

```json
"build:release": "npm-run-all build build:test build:api build:docs build:demoScripts bundler postinstall",
```

or if `build:demoScripts ` is not supported in that repo

```json
"build:release": "npm-run-all build build:test build:api build:docs bundler types postinstall",
```

I am suggesting removing `test` and `linters` from the release script as this would have ideally been run through multiple times by now. Also, this will be triggered with the pre-commit happening in the remote on the release pipeline. 

Lastly, update `build:ci` to reference the new `build:release` script. 

```json
"build:ci": "npm-run-all sweep build:release",
```

An updated `testPublish.yml` config proposal is then: 

```yml
name: Test and publish

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the main branch
on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  test:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x, 20.x]

    steps:
      - uses: actions/checkout@v4
      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ matrix.node-version }}
      - run: npm ci
      - run: npm run build
      - run: npm run build:test

  release:
    # Only release on push to main
    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
    runs-on: ubuntu-latest
    needs: test
    steps:
      - uses: actions/checkout@v4
        with:
          persist-credentials: false
      - uses: actions/setup-node@v4
        with:
          node-version: 20
      - run: npm ci
      - run: npm run build:release
      - uses: cycjimmy/semantic-release-action@v4
        env:
          GITHUB_TOKEN: ${{ secrets.ACCESS_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}

```

**Reply by blackfalcon** [reply]
This process was applied to recent updates of the following. Please be sure to update these repos if there are any changes to this recommendation. 

1. [auro-hyperlink](https://github.com/AlaskaAirlines/auro-hyperlink/pull/203/commits/11af03cd62ccb0c5a800d26c16816e2f581d8ae4) 
2. [auro-card](https://github.com/AlaskaAirlines/auro-card/pull/56/commits/0cd653963226e56e5bf3638699c4795de1e0ec9f)
3. [auro-popover](https://github.com/AlaskaAirlines/auro-popover/commit/27f08550086bc1c3a13ff03b0e9b86df77e1db51)
4. [auro-datepicker](https://github.com/AlaskaAirlines/auro-datepicker/commit/599758ea68aceb82d4beee79eb37ab04a66260a9)
5. [auro-dropdown](https://github.com/AlaskaAirlines/auro-dropdown/pull/229/commits/1214f1385f6d289b88deb6190034ee8eebcee26a)
6. [auro-input](https://github.com/AlaskaAirlines/auro-input/commit/f15200ecb7b75f2b37c2e66f902d7152f2a61300)

**Reply by jason-capsule42** [reply]
The proposed changes seem reasonable to me. I agree with implementing this as a standard for all repos.

**Comment by blackfalcon** [comment]
List of completed repos. 

|Repo|Updated|commit
|---|---|---
|wcss| [ ] | 
|icons| [ ] | 
|WC-Generator| [x] | https://github.com/AlaskaAirlines/WC-Generator/commit/4ef4d52534e82e0c966457ac0c7d457aed205a99
|auro-accordion| [x] | https://github.com/AlaskaAirlines/auro-accordion/commit/3ac51f6ee1fc2e76a25d08e1cea49acff3234237
|auro-alert| [x] | https://github.com/AlaskaAirlines/auro-alert/commit/a9e8a1699086480f8acf6fe2c24cf27b31835d3d
|auro-avatar| [x] | https://github.com/AlaskaAirlines/auro-avatar/commit/619bf81ecd1fb64e7d5331326630ebf27f15df59
|auro-background| [ ] | 
|auro-backtotop| [x] | https://github.com/AlaskaAirlines/auro-backtotop/commit/6999d5bbb5e0bf9d5a0db8f1e43ba9fcd33e0265
|auro-badge| [x] | https://github.com/AlaskaAirlines/auro-badge/commit/a19b9605cacb468035c43ef795929deab638e114
|auro-banner| [ ] | 
|auro-button| [x] | https://github.com/AlaskaAirlines/auro-button/commit/879b4530f4685e708c3a1c554d17d7b1a3a7a2d3
|auro-card| [x] | https://github.com/AlaskaAirlines/auro-card/pull/56/commits/0cd653963226e56e5bf3638699c4795de1e0ec9f
|auro-carousel| [x] | https://github.com/AlaskaAirlines/auro-carousel/commit/1f4d9dc5cc378186aadda03f2c39b135cbd2556a
|auro-checkbox| [x] | https://github.com/AlaskaAirlines/auro-checkbox/commit/7d4a92acf037b33d6ccb1676e3fbff3f7d2e8cc9
|auro-combobox| [x] | https://github.com/AlaskaAirlines/auro-combobox/commit/47f543908717917fa8b74cce1d963c5b6ebb8353
|auro-datetime| [x] | https://github.com/AlaskaAirlines/auro-datetime/commit/f0e1c4c7d1bddcc8edf52b708314e06ce1885e48
|auro-datepicker| [x] | https://github.com/AlaskaAirlines/auro-datepicker/commit/599758ea68aceb82d4beee79eb37ab04a66260a9
|auro-dialog| [x] | https://github.com/AlaskaAirlines/auro-dialog/commit/42fbdd3c238fc1162d069d580ff76133974c455c
|auro-drawer| [x] | https://github.com/AlaskaAirlines/auro-drawer/commit/45c68a2bc1965450ce30f8f15b142d15bbe39307
|auro-dropdown| [x] | https://github.com/AlaskaAirlines/auro-dropdown/commit/9a22ee7959327542c9c79f3ae8c2fc8ecde4c307
|auro-formvaliator| [ ] | n/a
|auro-flight| [ ] | https://github.com/AlaskaAirlines/auro-flight/commit/2a71c973bbeaaa605fb5af98713499540d210836
|auro-flightline| [x] | https://github.com/AlaskaAirlines/auro-flightline/commit/d97956ed7a31ab8164103e603c27ac33f0ec7f62
|auro-header| [x] | https://github.com/AlaskaAirlines/auro-header/commit/efff485f1a8ee64497b1967618705e18c0cd6b33
|auro-hyperlink| [x] | https://github.com/AlaskaAirlines/auro-hyperlink/pull/203/commits/11af03cd62ccb0c5a800d26c16816e2f581d8ae4
|auro-icon| [x] | https://github.com/AlaskaAirlines/auro-icon/pull/116/commits/ec7d2417721ac48d2650553a6faba5bc7ca7d1c0
|auro-input| [x] | https://github.com/AlaskaAirlines/auro-input/commit/f15200ecb7b75f2b37c2e66f902d7152f2a61300
|auro-loader| [x] | https://github.com/AlaskaAirlines/auro-loader/commit/e7b988585a2e2babaa52378aca54f36e80f5a4d5
|auro-lockup| [x] |https://github.com/AlaskaAirlines/auro-lockup/commit/08673fd09d21571fa305029d6c39f0d3520227f8
|auro-menu| [x] | https://github.com/AlaskaAirlines/auro-menu/commit/019cf053971396d4c372bd45e0885fd20d0ff87d
|auro-nav| [x] |   https://github.com/AlaskaAirlines/auro-nav/commit/0e37d2fe1400a4e11c89ede81908df94720dbc51
|auro-pane| [x] | https://github.com/AlaskaAirlines/auro-pane/commit/5299e523c08ba478cf6888d402e761e7fae40974
|auro-popover| [x] | https://github.com/AlaskaAirlines/auro-popover/commit/27f08550086bc1c3a13ff03b0e9b86df77e1db51
|auro-radio| [x] | https://github.com/AlaskaAirlines/auro-radio/commit/0f4ac88e809b238e0ce0bd189987d653c260a34f
|auro-select| [x] | https://github.com/AlaskaAirlines/auro-select/pull/190/commits/c6089e8a34f88eb93f8405281906385335e3aea5
|auro-skeleton| [x] | https://github.com/AlaskaAirlines/auro-skeleton/commit/2bccd761ccb13c4ce519dfe9e549345f06fba352
|auro-sidenav| [x] | https://github.com/AlaskaAirlines/auro-sidenav/commit/f2439a4a4993fe64b4701d1f49c530e8dd8089d4
|auro-table| [x] | https://github.com/AlaskaAirlines/auro-table/commit/7fd168654f9986d7d9a6ea0ec15a896ba1fce78c
|auro-toast| [x] | https://github.com/AlaskaAirlines/auro-toast/commit/c260820420b37d26937c9c42f8bd58f6b26b8586


**Script for commit:** 

build: update build scripts

This commit updates the pakcage.json build scripts to be compliant with
the decisions made in this discussion:

https://github.com/orgs/AlaskaAirlines/discussions/523
